### DISCUSSION TITLE
What is the need to declare classes final?

### DISCUSSION THREAD

**Discussion by greendrake** [reactphp/reactphp#495] [Q&A]
I find it pretty inconvenient that classes like `HttpServer` and `Io\StreamingServer` are `final`.

Say I want to extend the latter so that it emits an event after finishing outputting response to the client — so that I could bind some custom handler to that event in my app.

What I have to do now is to create a custom version of the `Http` package and rewrite those classes.

Shouldn't there be an easier, less code-repeating way?

**Comment by SimonFrings** [comment]
@greendrake Thanks for bringing this up :+1: 

I can understand that this might be a inconvenience, but there are reasons to mark these classes as final, for example to quote @clue's comment from https://github.com/reactphp/socket/pull/71:
> This reduces our API footprint and avoids future BC breaks by avoiding exposing its internal assumptions.

Not marking these classes as final would result in constant BC breaks once we change our behavior. Instead of using your custom version of the `http` package you can also use middleware in order to add your custom behavior.

For reference: https://github.com/reactphp/http/pull/283 and https://github.com/reactphp/socket/pull/71
